src/kivi/ast/src/ast/parsing_context.hh

```diff
@@ -0,0 +1,185 @@
+#ifndef KIVI_SRC_KIVI_PARSING_CONTEXT_HH_
+#define KIVI_SRC_KIVI_PARSING_CONTEXT_HH_
+
+/**
+ * @file parsing_context.hh
+ *
+ * This class contains all information required and used during the process
+ * of parsing a source file.
+ */
+
+#include <vector>
+#include <map>
+
+#define LOGURU_WITH_STREAMS 1
+#include <loguru/loguru.hpp>
+
+#include <kivi_parser/location.hh>
+#include <kivi_parser/stack.hh>
+
+#include <kivi_expressions/identifier.hh>
+#include <kivi_stmts/statement.hh>
+
+#include "syntax.hh"
+
+namespace syntax_analyzer
+{
+
+	class parsing_context
+	{
+	 private:
+		/// All parsed scopes
+		std::vector<std::map<std::string, identifier>> m_all_scopes;
+
+		/// All parsed functions
+		std::vector<function> m_all_functions;
+
+		/// The current function that is being processed
+		function m_this_function;
+
+		/// The current number of "registers" being used
+		int m_registers = 0;
+
+		/**
+		 * NB: The following 2 fields are public, because they are required by bison and re2c in order to keep track
+		 * of the current location of the concrete process being executed. Because they are publicly visible, they do
+		 * not follow the standard convention for field names - no "m_" prefix, and can be used without getters.
+		 */
+	 public:
+		/// The location which is used and updated by the re2c module
+		const char* lexer_cursor;
+
+		/// The location which is used and updated by the bison module
+		yy::location yy_location;
+
+	 public:
+
+		//
+		// Constructors
+		//
+
+		/**
+		 * The constructor initializes the "location" variables in both the
+		 * lexer and the parser.
+		 *
+		 * @param code_beginning  the "location" var for the lexer
+		 * @param filename the "location" var for the parser
+		 */
+		parsing_context(const char* code_beginning, std::string* filename)
```

Comment:
I do not think `explicit` would be a good addition, since the constructor accepts multiple arguments and I do not think implicit conversion could happen.

Reply:
There should be a way, somehow.. Let's put it in the backlog as a nice-to-have :wink: